src/python/pants/backend/codegen/protobuf/scala/rules.py

```diff
@@ -170,6 +225,128 @@ async def inject_scalapb_dependencies(
     return InjectedDependencies(addresses)
 
 
+@rule
+async def materialize_jvm_plugin(request: MaterializeJvmPluginRequest) -> MaterializedJvmPlugin:
+    requirements = await Get(
+        ArtifactRequirements,
+        GatherJvmCoordinatesRequest(
+            artifact_inputs=FrozenOrderedSet([request.plugin.artifact]),
+            option_name="--scalapb-plugin-artifacts",
+        ),
+    )
+    classpath = await Get(
+        MaterializedClasspath, MaterializedClasspathRequest(artifact_requirements=(requirements,))
+    )
+    return MaterializedJvmPlugin(
+        name=request.plugin.name,
+        classpath=classpath,
+    )
+
+
+SHIM_SCALA_VERSION = "2.13.7"
+
+
+@rule
+async def setup_scalapb_shim_classfiles(
+    scalapb: ScalaPBSubsystem, jdk_setup: JdkSetup, bash: BashBinary
+) -> ScalaPBShimCompiledClassfiles:
+    dest_dir = "classfiles"
+
```

Comment:
Would you mind opening a ticket about generalizing these bootstrap compilations (perhaps via synthetic targets and implicit dependencies... but even before that, just extracting shared `@rules` for it would be good), and linking it from the various locations where we're doing it?

Reply:
Will do as a separate PR (to include the other usage).